File: src/matreshka.py
import requests
from random import sample

class Matreshka:
	def __init__(self) -> None:
		self.api = "https://moblauncher.matrp.ru"
		self.headers = {
			"user-agent": "Dalvik/2.1.0 (Linux; U; Android 9; RMX3551 Build/PQ3A.190705.003)"
		}
		self.client = "prod"
		self.account_id = -1
		self.session_hash = None
		self.device_hash = self.generate_device_hash()

	def generate_device_hash(self) -> str:
		return "".join(sample("abcdefghijklmnopqrstuvwxyz" + "0123456789", 16))

	def login(
			self,
			email: str,
			password: str) -> dict:
		response = requests.get(
			f"{self.api}/api/Requests/Auth.php?type=email&subtype=auth&action=continue&email={email}&pass={password}&accountId={self.account_id}&sessionHash={self.session_hash}&deviceHash={self.device_hash}&client={self.client}",
			headers=self.headers).json()
		if "account_id" in response["actions"]:
			self.account_id = response["actions"]["account_id"]
			self.session_hash = response["actions"]["session_hash"]
		return response

	def request_verification_code(
			self,
			email: str,
			password: str) -> dict:
		return requests.get(
			f"{self.api}/api/Requests/Auth.php?type=email&subtype=register&action=continue&email={email}&pass={password}&repeat_pass={password}&accountId={self.account_id}&sessionHash={self.session_hash}&deviceHash={self.device_hash}&client={self.client}",
			headers=self.headers).json()

	def register(
			self,
			email: str,
			password: str,
			verification_code: int = None) -> dict:
		return requests.get(
			f"{self.api}/api/Requests/Auth.php?type=email&subtype=register_code&action=continue&email={email}&pass={password}&repeat_pass={password}&code={verification_code}&accountId={self.account_id}&sessionHash={self.session_hash}&deviceHash={self.device_hash}&client={self.client}",
			headers=self.headers).json()

	def get_stories(self) -> dict:
		return requests.get(
			f"{self.api}/api/Requests/Stories.php?accountId={self.account_id}&sessionHash={self.session_hash}&deviceHash={self.device_hash}&client={self.client}",
			headers=self.headers).json()

	def get_servers(self) -> dict:
		return requests.get(
			f"{self.api}/api/Requests/Servers.php?accountId={self.account_id}&sessionHash={self.session_hash}&deviceHash={self.device_hash}&client={self.client}",
			headers=self.headers).json()

	def get_social_bonus_info(self) -> dict:
		return requests.get(
			f"{self.api}/api/Requests/SocialBonusInfo.php?accountId={self.account_id}&sessionHash={self.session_hash}&deviceHash={self.device_hash}&client={self.client}",
			headers=self.headers).json()

	def get_account_details(self) -> dict:
		return requests.get(
			f"{self.api}/api/Requests/AccountDetails.php?accountId={self.account_id}&sessionHash={self.session_hash}&deviceHash={self.device_hash}&client={self.client}",
			headers=self.headers).json()

	def get_cabinet(self) -> dict:
		return requests.get(
			f"{self.api}/api/Requests/Cabinet.php?accountId={self.account_id}&sessionHash={self.session_hash}&deviceHash={self.device_hash}&client={self.client}",
			headers=self.headers).json()
	
	def change_password(
			self,
			password: str) -> dict:
		return requests.get(
			f"{self.api}/api/Requests/Auth.php?type=email&subtype=changepass_old&action=continue&pass={password}&accountId={self.account_id}&sessionHash={self.session_hash}&deviceHash={self.device_hash}&client={self.client}",
			headers=self.headers).json()

	def get_faq(self) -> dict:
		return requests.get(
			f"{self.api}/api/Requests/FAQ.php?accountId={self.account_id}&sessionHash={self.session_hash}&deviceHash={self.device_hash}&client={self.client}",
			headers=self.headers).json()

	def get_files(self) -> dict:
		return requests.get(
			f"{self.api}/files_list.php?accountId={self.account_id}&sessionHash={self.session_hash}&deviceHash={self.device_hash}&client={self.client}",
			headers=self.headers).json()

	def get_audios(self) -> dict:
		return requests.get(
			f"{self.api}/audio_list.php?accountId={self.account_id}&sessionHash={self.session_hash}&deviceHash={self.device_hash}&client={self.client}",
			headers=self.headers).json()

	def get_radios(self) -> dict:
		return requests.get(
			f"{self.api}/radio_list.php?accountId={self.account_id}&sessionHash={self.session_hash}&deviceHash={self.device_hash}&client={self.client}",
			headers=self.headers).json()

	def get_update(self) -> dict:
		return requests.get(
			f"{self.api}/update.php?accountId={self.account_id}&sessionHash={self.session_hash}&deviceHash={self.device_hash}&client={self.client}",
			headers=self.headers).json()

File: src/matreshka.js
class Matreshka {
	constructor(countryCode = "ru") {
		this.api = "https://moblauncher.matrp.ru"
		this.headers = {
			"user-agent": "Dalvik/2.1.0 (Linux; U; Android 9; RMX3551 Build/PQ3A.190705.003)"
		}
		this.client = "prod"
		this.accountId = -1
		this.deviceHash = this.generateDeviceHash
	}

	generateDeviceHash() {
		let hash = ""
		const characters = "abcdefghijklmnopqrstuvwxyz0123456789"
		for (let i = 0; i < 16; i++) {
			hash += characters.charAt(Math.floor(Math.random() * characters.length));
		}
		return hash
	}

	async login(email, password) {
		const response = await fetch(
			`${this.api}/api/Requests/Auth.php?type=email&subtype=auth&action=continue&email=${email}&pass=${password}&accountId=${this.accountId}&sessionHash=${this.sessionHash}&deviceHash=${this.deviceHash}&client=${this.client}`, {
				method: "GET",
				headers: this.headers
			})
		const data = await response.json()
		if ("account_id" in data.actions[0]) {
			this.accountId = data.actions[0].account_id
			this.sessionHash = data.actions[0].session_hash
		}
		return data
	}

	async requestVerificationCode(email, password) {
		const response = await fetch(
			`${this.api}/api/Requests/Auth.php?type=email&subtype=register&action=continue&email=${email}&pass=${password}&repeat_pass=${password}&accountId=${this.accountId}&sessionHash=${this.sessionHash}&deviceHash=${this.deviceHash}&client=${this.client}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async register(email, password, verificationCode) {
		const response = await fetch(
			`${this.api}/api/Requests/Auth.php?type=email&subtype=register_code&action=continue&email=${email}&pass=${password}&repeat_pass=${password}&code=${verificationCode}&accountId=${this.accountId}&sessionHash=${this.sessionHash}&deviceHash=${this.deviceHash}&client=${this.client}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getStories() {
		const response = await fetch(
			`${this.api}/api/Requests/Stories.php?accountId=${this.accountId}&sessionHash=${this.sessionHash}&deviceHash=${this.deviceHash}&client=${this.client}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getServers() {
		const response = await fetch(
			`${this.api}/api/Requests/Servers.php?accountId=${this.accountId}&sessionHash=${this.sessionHash}&deviceHash=${this.deviceHash}&client=${this.client}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getSocialBonusInfo() {
		const response = await fetch(
			`${this.api}/api/Requests/SocialBonusInfo.php?accountId=${this.accountId}&sessionHash=${this.sessionHash}&deviceHash=${this.deviceHash}&client=${this.client}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getAccountDetails() {
		const response = await fetch(
			`${this.api}/api/Requests/AccountDetails.php?accountId=${this.accountId}&sessionHash=${this.sessionHash}&deviceHash=${this.deviceHash}&client=${this.client}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getCabinet() {
		const response = await fetch(
			`${this.api}/api/Requests/Cabinet.php?accountId=${this.accountId}&sessionHash=${this.sessionHash}&deviceHash=${this.deviceHash}&client=${this.client}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async changePassword(password) {
		const response = await fetch(
			`${this.api}/api/Requests/Auth.php?type=email&subtype=changepass_old&action=continue&pass=${password}&accountId=${this.accountId}&sessionHash=${this.sessionHash}&deviceHash=${this.deviceHash}&client=${this.client}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getFaq() {
		const response = await fetch(
			`${this.api}/api/Requests/FAQ.php?accountId=${this.accountId}&sessionHash=${this.sessionHash}&deviceHash=${this.deviceHash}&client=${this.client}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getAudios() {
		const response = await fetch(
			`${this.api}/audio_list.php?accountId=${this.accountId}&sessionHash=${this.sessionHash}&deviceHash=${this.deviceHash}&client=${this.client}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getRadios() {
		const response = await fetch(
			`${this.api}/radio_list.php?accountId=${this.accountId}&sessionHash=${this.sessionHash}&deviceHash=${this.deviceHash}&client=${this.client}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getUpdate() {
		const response = await fetch(
			`${this.api}/update.php?accountId=${this.accountId}&sessionHash=${this.sessionHash}&deviceHash=${this.deviceHash}&client=${this.client}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}
}

module.exports = {Matreshka}
